cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionFilter.java

```diff
@@ -0,0 +1,245 @@
+     *         or if the value is null.
+     */
+    public <T extends Comparable<T>> Builder addValueCondition(String field, T value) {
+      Preconditions.checkArgument(field != null && !field.isEmpty(), "field name cannot be null or empty.");
+      Preconditions.checkArgument(value != null, "condition value cannot be null.");
+      if (map.containsKey(field)) {
+        throw new IllegalArgumentException(String.format("Field '%s' already exists in partition filter.", field));
+      }
+      map.put(field, new Condition<T>(field, value));
+      return this;
+    }
+
+    /**
+     * Create the PartitionFilter.
+     *
+     * @throws java.lang.IllegalStateException if no fields have been added
+     */
+    public PartitionFilter build() {
+      Preconditions.checkState(!map.isEmpty(), "Partition filter cannot be empty.");
+      return new PartitionFilter(map);
+    }
+  }
+
+  /**
+   * Represents a condition on a partitioning field, by means of an inclusive lower bound and an exclusive upper bound.
+   * As a special case, if only one value is passed to the constructor, then this represents an equality filter.
+   * @param <T> The type of the partitioning field.
+   */
+  public static class Condition<T extends Comparable<T>> {
+
+    private String fieldName;
+    private final T lower;
+    private final T upper;
+    private final boolean isSingleValue;
+
+    private Condition(String fieldName, T lower, T upper) {
+      Preconditions.checkArgument(lower != null || upper != null, "Either lower or upper-bound must be non-null.");
+      this.fieldName = fieldName;
+      this.lower = lower;
+      this.upper = upper;
+      this.isSingleValue = false;
+    }
+
+    private Condition(String fieldName, T value) {
+      Preconditions.checkArgument(value != null, "Value must not be null.");
+      this.fieldName = fieldName;
+      this.lower = value;
+      this.upper = null;
+      this.isSingleValue = true;
+    }
+
+    /**
+     * @return the expected value of this condition
+     */
+    public T getValue() {
+      return lower;
+    }
+
+    /**
+     * @return the lower bound of this condition
+     */
+    public T getLower() {
+      return lower;
+    }
+
+    /**
+     * @return the upper bound of this condition
+     */
+    public T getUpper() {
+      return upper;
+    }
+
+    /**
+     * @return whether this condition matches a single value
+     */
+    public boolean isSingleValue() {
+      return isSingleValue;
+    }
+
+    /**
+     * Match the condition against a given value. The value must be of the same type as the bounds of the condition.
+     *
+     * @throws java.lang.IllegalArgumentException if the value has an incompatible type
+     */
+    public <V extends Comparable> boolean match(V value) {
+      try {
+        // if lower and upper are identical, then this represents an equality condition.
+        if (isSingleValue()) {
+          return getValue().equals(value);
```

Comment:
Since you require `value` as `Comparable, would it be better to compare with `getValue().compareTo(value) == 0`?